kythe/go/localrun/localrun.go

```diff
@@ -0,0 +1,518 @@
+
+	rd, err := dedup.NewReader(r.indexedOut, int(r.CacheSize.Bytes()))
+	if err != nil {
+		return fmt.Errorf("error creating deduped stream: %v", err)
+	}
+
+	tmpfile, err := ioutil.TempFile("", "dedup_kythe_stream.pb")
+	if err != nil {
+		return fmt.Errorf("error creating tmpfile: %v", err)
+	}
+	defer os.Remove(tmpfile.Name())
+
+	wr := delimited.NewWriter(tmpfile)
+	if err := delimited.Copy(wr, rd); err != nil {
+		return fmt.Errorf("unable to copy deduped stream to disk: %v", err)
+	}
+
+	p, s := beam.NewPipelineWithRoot()
+	entries, err := beamio.ReadEntries(ctx, s, tmpfile.Name())
+	if err != nil {
+		return fmt.Errorf("error reading entries: %v", err)
+	}
+	k := pipeline.FromEntries(s, entries)
+
+	beamio.WriteLevelDB(s, r.OutputDir, r.WorkerPoolSize,
+		createColumnarMetadata(s),
+		k.SplitCrossReferences(),
+		k.SplitDecorations(),
+		k.CorpusRoots(),
+		k.Directories(),
+		k.Documents(),
+		k.SplitEdges(),
+	)
+
+	return beamx.Run(ctx, p)
+
+	// dedup the stream
+	// https://github.com/kythe/kythe/blob/77e1360beca7903bc97728097d2c5e1b82204092/kythe/go/platform/tools/dedup_stream/dedup_stream.go
+	// https://github.com/kythe/kythe/blob/77e1360beca7903bc97728097d2c5e1b82204092/kythe/go/serving/tools/write_tables/write_tables.go#L139
+}
+
+func createColumnarMetadata(s beam.Scope) beam.PCollection {
+	return beam.ParDo(s, emitColumnarMetadata, beam.Impulse(s))
+}
+
+func emitColumnarMetadata(_ []byte) (string, string) { return xrefs.ColumnarTableKeyMarker, "v1" }
+
+// Serve starts an http server on the provided port and allows web inspection
+// of the indexed data.
+func (r *Runner) Serve(ctx context.Context) error {
+	if err := r.checkMode(serve); err != nil {
+		return err
+	}
+
+	listen := fmt.Sprintf("%s:%d", r.Hostname, r.Port)
+
+	args := []string{
+		"--listen", listen,
+		"--serving_table", r.OutputDir,
+	}
+
+	if _, err := os.Stat(r.PublicResources); !os.IsNotExist(err) {
+		args = append(args, "--public_resources")
+		args = append(args, r.PublicResources)
+	} else if r.PublicResources != "" {
+		log.Printf("You requested serving public resources from %q, but it doesn't exist.", r.PublicResources)
+	}
+
+	// Perform the work prescribed
+	cmd := exec.CommandContext(ctx, fmt.Sprintf("%s/tools/http_server", r.KytheRelease), args...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Dir = r.WorkingDir
+
+	log.Printf("Starting http server on http://%s", listen)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("error starting serve: %w", err)
+	}
+	return nil
+}
+
+type indexer interface {
+	run(ctx context.Context, kzips []string) error
+}
+
+type serialIndexer struct {
+	besFile      string
+	Languages    LanguageSet
+	KytheRelease string
+	WorkingDir   string
+	Timeout      time.Duration
+}
+
+func (si *serialIndexer) run(ctx context.Context, kzips []string) (io.Reader, error) {
+	indexedOut := &bytes.Buffer{}
+
+	log.Printf("Beginning serial indexing")
+
+	for i, k := range kzips {
```

Comment:
The output should probably be deduped as well.  A TODO can be added for that.

Reply:
Added a TODO. Out of curiosity, what would you use as the deduplication criteria?

Reply:
I meant dedup the entry stream output as it is done when feeding it into the post-processor.  Doing it sooner means less disk IO/space used during indexing.